File: src/part02/chapter10/levelA/task11/java/logic/PolynomialHelper.java
package part02.chapter10.levelA.task11.java.logic;

import java.util.HashMap;

public class PolynomialHelper {

    public HashMap<Integer, Double> addPolynomials(
            HashMap<Integer, Double> firstPolynomialsCoefficients,
            HashMap<Integer, Double> secondPolynomialsCoefficients
    ) {
        if (firstPolynomialsCoefficients.size() != secondPolynomialsCoefficients.size()) {
            throw new IllegalArgumentException();
        }
        HashMap<Integer, Double> result = new HashMap<>();
        for (int i = 0; i < firstPolynomialsCoefficients.size(); i++) {
            double currentCoefficient = firstPolynomialsCoefficients.get(i) + secondPolynomialsCoefficients.get(i);
            result.put(i, currentCoefficient);
        }
        return result;
    }

    public double obtainValueOfPolynomial(double variable, HashMap<Integer, Double> coefficients) {
        double result = 0;
        for (int i = 0; i < coefficients.size(); i++) {
            result += coefficients.get(i) * Math.pow(variable, coefficients.size() - i - 1);
        }
        return result;
    }

    public void printPolynomial(double variable, HashMap<Integer, Double> coefficients) {
        for (int i = 0; i < coefficients.size(); i++) {
            if (i < coefficients.size() - 1) {
                System.out.println(coefficients.get(i) + " * (" + variable + ")^" + (coefficients.size() - i - 1) + "\t+");
            } else {
                System.out.println(coefficients.get(i) + " * (" + variable + ")^" + (coefficients.size() - i - 1));
            }
        }
    }
}

File: src/part02/chapter10/levelA/task11/kotlin/logic/PolynomialHelper.kt
package part02.chapter10.levelA.task11.kotlin.logic

import kotlin.math.pow

object PolynomialHelper {

    fun addPolynomials(firstCoefficients: HashMap<Int, Double>, secondCoefficients: HashMap<Int, Double>): HashMap<Int, Double> {
        if (firstCoefficients.size != secondCoefficients.size) throw IllegalArgumentException()
        val countOfCoefficients = firstCoefficients.size
        val result = HashMap<Int, Double>(countOfCoefficients)
        for (i in 0 until  countOfCoefficients) {
            val firstCoefficient = firstCoefficients[i]
            val secondCoefficient = secondCoefficients[i]
            if (firstCoefficient != null && secondCoefficient != null) {
                result[i] = firstCoefficient + secondCoefficient
            } else {
                throw IllegalArgumentException()
            }
        }
        return result
    }

    fun obtainValueOfPolynomial(variable: Double, coefficients: HashMap<Int, Double>):Double {
        var result = 0.0
        for (i in 0 until coefficients.size) {
            val coefficient = coefficients[i] ?: throw IllegalArgumentException()
            result += coefficient * variable.pow(i)
        }
        return result
    }

    fun printPolynomial(variable: Double, coefficients: HashMap<Int, Double>) {
        for (i in 0 until coefficients.size) {
            if (i != coefficients.size - 1) {
                println("${coefficients[i]} * ($variable)^${coefficients.size - 1 -i}\t+")
            } else {
                println("${coefficients[i]} * ($variable)^${coefficients.size - 1 -i}")
            }
        }
    }
}